### ISSUE TITLE
Unable to access `get_info` without `DeprecationWarning`

### ISSUE THREAD

**Issue by bashtage** [numpy/numpy#21431]
### Describe the issue:

`get_info` is needed to locate various libraries and headers for building extension modules. It is located in `from numpy.distutils.misc_util`, and importing it seems to lead to a `DeprecationWarning`

### Reproduce the code example:

```python
from numpy.distutils.misc_util import get_info
```

### Error message:

```shell
`numpy.distutils` is deprecated since NumPy 1.23.0, as a result
    of the deprecation of `distutils` itself. It will be removed for
    Python >= 3.12. For older Python versions it will remain present.
    It is recommended to use `setuptools < 60.0` for those Python versions.
    For more details, see:
      https://numpy.org/devdocs/reference/distutils_status_migration.html
```
```


### NumPy/Python version information:

From CI so not easy to run command directly

```
Python 3.10.4 x64 (Linux)
1.23.0.dev0+1087.g0eaa40db3
```

**Comment by charris** [comment]
Hmm. NumPy distutils is deprecated as a whole, but perhaps we will need something like `get_info` going forward. Do you use other parts of distutils, or just that function? @rgommers I assume that the SciPy meson build has equivalent functionality somewhere.

**Comment by bashtage** [comment]
Only `get_info` is really needed.  We also import `set_verbosity` but that can be ignored.  The actual build is standard. (referring to statsmodels)

**Comment by charris** [comment]
`get_info` uses `sys_info`, which imports a _lot_ of stuff from distutils, so moving the function somewhere else is non-trivial. Can be done if there is no other solution to the problem.

**Comment by rgommers** [comment]
It is actually deprecated. The design is bad, so I don't plan to migrate that API. Meson uses `pkg-config` for dependency discovery (in addition it can also have custom lookups if needed), but anyway it doesn't make sense to do dependency discovery in isolation.

For libraries that do need this and want to stay with `setuptools`, we could move `get_info` into `setuptools`. 

@bashtage you only need it for `npymath`, right? In https://github.com/statsmodels/statsmodels/blob/23a34f502e2c6bad5d7a6bec34160917b7ea541b/setup.py#L217-L221

I recommend simply using `path_to_npymath = os.path.join(np.get_include(), '..', 'lib')`. That's what I did in the SciPy build as well, also for Meson (until Meson gets native knowledge of NumPy).

**Comment by rgommers** [comment]
To expand a little, what stays is:
- `numpy.get_include`
- `numpy.f2py.get_include`

For `npymath` and `npyrandom` there's no API call, but the location can be derived from `numpy.get_include`. We _could_ add a new API that's more generic and covers all numpy-internal include and library dirs. It'd just be a minor convenience at this point, but probably a good idea - it's also why I added `f2py.get_include` recently.

For external dependencies that `numpy/distutils/system_info.py` covers - that all should die preferably. It's badly designed and the hardcoded paths break cross-compilation, and sometimes even regular builds (e.g. with conda, since it uses `-Isysroot`) in subtle ways. External dependency lookup is a thing you want from your build system. If you're using `numpy.distutils` and are needing `system_info` for BLAS, FFTW or similar: I recommend migration to Meson.

**Comment by rgommers** [comment]
Relabeled to a new feature we should consider adding for 1.24.0

**Comment by rgommers** [comment]
Of course I'm not yet sure that adding this new feature is the right thing to do, since it's still not usable for cross-compiling if you can't do `python -c import numpy` for the host machine (as opposed to the build machine). So using `pkg-config` is a better idea anyway.

**Comment by bashtage** [comment]
@rgommers Thanks, that work around should help.  I suspect that `get_info` is widely used by libraries that link to NumPy, so it would be good to have a replacement or a clearly documented path forward somewhere.

**Comment by rgommers** [comment]
I'd actually be surprised if it's widely used. `npymath` is a likely reason; let me also xref gh-20880 for a related `npymath` design/packaging issue.

For external dependencies, I expect it's only a handful of BLAS/LAPACK and FFTW usages. Everything else is pretty much obsolete.

**Comment by excubo-jg** [comment]
> @rgommers Thanks, that work around should help. I suspect that `get_info` is widely used by libraries that link to NumPy, so it would be good to have a replacement or a clearly documented path forward somewhere.

Is there some documentation available how to avoid the get_info reference? I am using this code from the documentation https://numpy.org/doc/stable/reference/random/examples/cython/setup.py.html and would like to remove the quite expansive warning message.

**Comment by rgommers** [comment]
@excubo-jg see gh-22245.

**Comment by excubo-jg** [comment]
Many thanks - removed the warning message at my end

**Comment by mattip** [comment]
@excubo-jg can we close this then?

**Comment by excubo-jg** [comment]
From my end yes - although I did not open the issue. Many thanks for the quick turnaround

**Comment by rgommers** [comment]
There is more to do here, let's leave it open. `get_info` will naturally be removed, but more importantly: we do need a replacement, probably by extending `np.get_include` to allow querying for `npymath` & co.

**Comment by seberg** [comment]
@rgommers how important is it that tis gets looked in for the NumPy 1.24 release that is upcoming?  If we want to add new API it would be good to push it soon.

**Comment by mattip** [comment]
> probably by extending `np.get_include` to allow querying for `npymath` & co

Maybe we should have a np.build namespace, and provide np.build.pkg_config() with kwargs that mimic the `pkg-config` ones

**Comment by rgommers** [comment]
An update on how this currently works. We write two pkg-config like files to `numpy/core/lib/npy-pkg-config`. First there is `mlib.ini`:
```ini
[meta]
Name = mlib
Description = Math library used with this version of numpy
Version = 1.0

[default]
Libs=-lm
Cflags=

[msvc]
Libs=m.lib
Cflags=
```

Second there is `npymath.ini`:
```ini
[meta]
Name=npymath
Description=Portable, core math library implementing C99 standard
Version=0.1

[variables]
pkgname=numpy.core
prefix=${pkgdir}
libdir=${prefix}/lib
includedir=${prefix}/include

[default]
Libs=-L${libdir} -lnpymath
Cflags=-I${includedir}
Requires=mlib

[msvc]
Libs=/LIBPATH:${libdir} npymath.lib
Cflags=/INCLUDE:${includedir}
Requires=mlib
```

There is no matching `npyrandom.ini`, all we ship is `numpy/random/lib/libnpyrandom.a`. `get_info('npyrandom')` fails, while for `npymath` we get:
```python
In [29]: get_info('npymath')
Out[29]: 
{'include_dirs': ['/home/rgommers/code/numpy/numpy/core/include'],
 'library_dirs': ['/home/rgommers/code/numpy/numpy/core/lib'],
 'libraries': ['npymath', 'm'],
 'define_macros': []}
```

The `mlib` thing doesn't look completely portable. It's not completely obvious what we should do here. `np.get_include` just returns the location of the headers, while `get_info` returns also the location of the static library and the link line that is needed. The options are:
1. Do nothing, document the current workaround
2. Move `get_info` into a new namespace, either `numpy` or `numpy.lib`
3. Extend `get_include` to also return library location and (maybe) linker flags

Given that `npymath` and `npyrandom` shipping as static libraries is problematic to begin with, and we have the plan to make `npymath` into a vendor-able source package instead, we may leave it at (1) for now.

`npyrandom` is much more problematic, it is not easy to rebuild from source / make vendor-able. The [docs for it](https://numpy.org/devdocs/reference/random/c-api.html) currently just state: _" To access these from Cython or C, you must link with the `npyrandom` library which is part of the NumPy distribution, located in `numpy/random/lib`."_ It doesn't even say that you must also link in `npymath` to make it work. Nor does it provide a header to go with the static library, unlike `npy_math.h`. Here is what's done in the `setup.py` of the extending in Cython example: https://github.com/numpy/numpy/blob/2081f2997b3ff192604415587761ce1806dc20a8/numpy/random/_examples/cython/setup.py. That is basically the same as the "do nothing" option here.

I'll open a PR to update the docs, and propose we leave it at that.

**Comment by rgommers** [comment]
done in gh-22538

**Comment by enzbus** [comment]
Hello; Python 3.12 breaks a few packages because their build systems used `numpy.distutils.get_info` to locate Blas/Lapack, get compile flags, etc. `numpy.system_info` is not a replacement apparently. Can you make a patch to bring just that function back? It shouldn't be too hard, and doesn't require you to change your own build system... Thanks!

**Comment by mattip** [comment]
What package are you trying to build with python 3.12 that still uses `get_info`? They should move away from that and use meson (which uses pkg-config underneath) or CMake instead. We are very unlikely to restore `numpy.distutils`.

**Comment by enzbus** [comment]
Thanks for replying! This is for [SCS](https://github.com/cvxgrp/scs), a solver for convex optimization problems. Its pip package uses `numpy.distutils.get_info` to get linkage information for Blas and Lapack, doesn't use anything else from `numpy.distutils`. ([See here](https://github.com/bodono/scs-python/issues/69).) This was buried in a Github CI job and we only realized it was broken when we tried to build Python 3.12 packages. I see that `numpy.system_info` gives you the location of the Blas/Lapack against which `numpy` is built. Is there any way we can also get the compilation flags and the other information that `numpy.distutils.get_info` provided? This is only for Blas and Lapack. Thanks!

**Comment by rgommers** [comment]
> Is there any way we can also get the compilation flags and the other information that numpy.distutils.get_info provided? This is only for Blas and Lapack.

I'm afraid not. One, because it's not easy to recreate that same output in the Meson build. And more importantly, because those flags should not be relevant. The way you were using them till now is very fragile, `numpy`'s build-time info has no guarantees at all to apply in the build environment of any other package.

> I see that `numpy.system_info` gives you the location of the Blas/Lapack against which `numpy` is built

It doesn't do that - instead, it actually goes and detect the BLAS/LAPACK on your system. Which may be the same as the ones `numpy` was built with, or it may be different.

I'll follow https://github.com/bodono/scs-python/issues/69 too, I think the discussion is best continued there.

**Comment by enzbus** [comment]
Thanks @rgommers , you are right. SCS is not my project but a few others (including [CVXPY](https://github.com/cvxpy/cvxpy), and another library I maintain) depend on it, so I was hoping to get a quick fix to deliver 3.12 packages. `distutils`' deprecation and the lack of a canonical packaging process has been problematic for many, unfortunately. We'll figure something out, if you have other inputs please share on https://github.com/bodono/scs-python/issues/69 . Thanks!
